apps/builder/app/builder/features/settings-panel/variables.tsx

```diff
@@ -0,0 +1,362 @@
+  if (instanceSelector === undefined) {
+    return;
+  }
+  const [instanceId] = instanceSelector;
+  serverSyncStore.createTransaction([dataSourcesStore], (dataSources) => {
+    let variableValue: VariableDataSource["value"] = {
+      type: "string",
+      value: "",
+    };
+    if (typeof value === "string") {
+      variableValue = { type: "string", value };
+    }
+    if (typeof value === "number") {
+      variableValue = { type: "number", value };
+    }
+    if (typeof value === "boolean") {
+      variableValue = { type: "boolean", value };
+    }
+    dataSources.set(dataSourceId, {
+      id: dataSourceId,
+      // preserve existing instance scope when edit
+      scopeInstanceId:
+        dataSources.get(dataSourceId)?.scopeInstanceId ?? instanceId,
+      name,
+      type: "variable",
+      value: variableValue,
+    });
+  });
+};
+
+const VariablePanel = ({
+  variable,
+  onBack,
+}: {
+  variable?: VariableDataSource;
+  onBack: () => void;
+}) => {
+  // variable value cannot have an access to other variables
+  const variables = useMemo(() => new Map(), []);
+  const nameId = useId();
+  const [name, setName] = useState(variable?.name ?? "");
+  const [nameErrors, setNameErrors] = useState<undefined | string[]>();
+  const [value, setValue] = useState(
+    JSON.stringify(variable?.value.value ?? "")
+  );
+  const [valueErrors, setValueErrors] = useState<undefined | string[]>();
+
+  return (
+    <ScrollArea
+      css={{
+        display: "flex",
+        flexDirection: "column",
+        width: theme.spacing[30],
+      }}
+    >
+      <Flex
+        direction="column"
+        css={{
+          overflow: "hidden",
+          p: theme.spacing[9],
+          gap: theme.spacing[5],
+        }}
+      >
+        <Flex direction="column" css={{ gap: theme.spacing[3] }}>
+          <Label htmlFor={nameId}>Name</Label>
+          <InputErrorsTooltip errors={nameErrors}>
+            <InputField
+              id={nameId}
+              value={name}
+              onChange={(event) => setName(event.target.value)}
+            />
+          </InputErrorsTooltip>
+        </Flex>
+        <Flex direction="column" css={{ gap: theme.spacing[3] }}>
+          <Label>Value</Label>
+          <InputErrorsTooltip errors={valueErrors}>
+            <div>
+              <CodeEditor
+                variables={variables}
+                defaultValue={value}
+                onChange={setValue}
+              />
+            </div>
+          </InputErrorsTooltip>
+        </Flex>
+        <Flex justify="end" css={{ gap: theme.spacing[5] }}>
+          <Button color="neutral" onClick={onBack}>
+            Cancel
+          </Button>
+          <Button
+            onClick={() => {
+              if (/[a-zA-z_]+[a-zA-z0-9_]*/.test(name) === false) {
+                setNameErrors([
+                  `Variable name can star with letter or _ and contain letters, digits and underscores`,
```

Comment:
Errors in text star/start, 
can - must
Shoudl be:
`Variable names must start with a letter or underscore (_) and can include numbers and underscores.`